### ISSUE TITLE
Exclude untracked files (pytest + coverage)

### ISSUE THREAD

**Issue by tilboerner** [pre-commit/pre-commit#708]
I'm currently trying out running pytest together with a coverage target as a pre-commit hook. This is problematic when there are untracked files present, because a commit should not be prevented by test failures or missing test coverage of untracked files (which are not getting committed, after all).

For this to work, and unless I'm missing something, it currently seems necessary to manually stash untracked files (exclusively) before `pre-commit run`, which comes down to doing something like: 

```
$ git stash --keep-index && \
git stash --keep-index --include-untracked && \
git stash pop stash@{1}
$ pre-commit run
$ git clean -f && git stash pop
```

Could pre-commit be made to do this for me and stash and restore untracked files, just like it's already hiding unstaged changes? This seems to me the cleanest way to have all hooks run in an environment that looks like the actual commit.

For what it's worth, I haven't found a good way to configure a `system` hook to work around unstaged files. As tests are not necessarily in the same file as the tested code, I think `pytest` cannot just run for files with staged changes: the test might be in an unchanged file. Also, the coverage plugin for pytest offers no command-line arguments for changing the files coverage looks at; therefore, `coverage` needs to be invoked directly. If I'm not mistaken, the resulting hook would somehow need to run this command (for an arbitrary coverage target of 99%):

```
$ coverage run -m pytest --ignore=$UNSTAGED_FILES && \
coverage report --omit=$UNSTAGED_FILES --fail-under=99
```

I could probably put this in a dedicated script. However, it would be much nicer if pre-commit took care of unstaged files for me.

*edit* In the above command, `$UNSTAGED_FILES` should be `$UNTRACKED_FILES`, of course. Sorry for the confusion.

**Comment by asottile** [comment]
So pre-commit already stashes unstaged changes, but not for untracked files (for pre-commit's filename collection's sake they're irrelevant as pre-commit _only_ considers tracked files).

I don't think mangling untracked files is feasible as it may blow away the pre-commit executable itself (A frequent setup is to use `pre-commit` via tox -- which is an untracked file)!  So the fix here should really be in the hook.

Most of pre-commit's design is around taking a list of changed files and passing them to an executable that knows how to process these -- running the whole testsuite ~kind of goes against this but it's definitely doable.  I tend to discourage this use of pre-commit also as it's intended to be lightning-fast checks when committing to do spot checks, and leave full testsuites for CI.  That said, this should be possible so let me see if I can't help you make this work :)

The trick here (in your hook) is to figure out what files aren't relevant (untracked test files) and exclude them from pytest and coverage.

You're right about `pytest-cov` -- it doesn't quite support the options you're looking for.  Personally, I don't use `pytest-cov` due to several edgecases (conftest, import-time, etc.) and usually invoke coverage in the way you showed above.  That said, we can still make this work with `pytest-cov` if that's your eventual goal by:

1. write out a temporary `.coveragerc` with `omit=...`
2. Pass `--cov-config=$temp_coveragerc`
3. remove `.coveragerc` (probably in a `finally` / `contextmanager` (or `TRAP` if you're going to go the `language: script` route)).

[This SO post](https://stackoverflow.com/q/3801321/812183) can probably assist with the finding of files?

Note also that `pre-commit` intentionally avoids using `git stash` as it could interfere with user workflows (and stash pop conflict sadness) and instead implements the unstaged stashing with `git diff-index`, `git checkout`, `git apply` [here](https://github.com/pre-commit/pre-commit/blob/1f13a5b6b917b10d3b90a35de3fb6ae03312d7bc/pre_commit/staged_files_only.py#L27-L30).

**Comment by tilboerner** [comment]
Thank you for your quick and very comprehensive answer! I concur with the philosophy of having very fast hooks; but the test "suite" I want to run here is over in the blink of an eye, and in this case it might as well be part of the pre-commit checks. The more gets caught before CI the better.

Good point about not disappearing the untracked files of the general public -- there are various reasons why people might not appreciate that.

Thanks as well for the pointers concerning implementation. Using a temporary .coveragerc would work, although, I think, at the expense of being able to keep the coverage configuration together with others in `setup.cfg`. Also, it would be nice if the hook could detect `--all-files` and *not* exclude untracked files in that case. 

But these are minor inconveniences; your suggestions should work. I might also decide that I can live with manually moving untracked files out of the way in the rare moments when it is necessary. Although I must say, it would be very handy to be able to automate it in a more robust way.

I still have the suspicion that "stash untracked" and `.gitignore` should be the obvious choices to handle a situation like this, and if they aren't, not in general, then there must be room for improvement in these tools themselves *or* how we're using them. But then again, that's not necessarily part of this ticket or even this project. :)

**Comment by asottile** [comment]
Cheers -- thanks again for the issue :)

**Comment by lorenzwalthert** [comment]
Sorry to jump in here again but I did not fully understand how untracked files should be handled in general when they affect whether or not the hook passes, i.e. I ran into the same situation as @tilboerner. I had to manually stash untracked files for pre-commit to work correctly because one of the untracked files affected other files in the repo when the hook of interest was running. For this reason, the hook was not passing. What is best practice here or am I just doing something wrong? Because @asottile says above that the untracked files are irrelevant for pre-commit anyways?

> for pre-commit's filename collection's sake they're irrelevant as pre-commit only considers tracked files.

Thanks.
